### PR TITLE
Assert slide parent is connected

### PR DESCRIPTION
**Pull request by choumx**
Follow-up to #9426.

/to @jridgewell @camelburrito

### CODE REVIEW

File: src/dom.js [hunk 1]
```diff
@@ -161,6 +161,38 @@ export function createElementWithAttributes(doc, tagName, attributes) {
...
+  let n = node;
+  do {
+    n = rootNodeFor(n);
+    if (n.host) {
+      n = n.host;
+    } else {
+      break;
+    }
+  } while (true);
+  return (n instanceof Document);
```
Comment (camelburrito) [inline src/dom.js]: how bad is this for performance - should we add guidelines of when/whom should use this? (may be add a presubmit review hook?)

File: src/dom.js [hunk 2]
```diff
@@ -161,6 +161,39 @@ export function createElementWithAttributes(doc, tagName, attributes) {
...
```
Comment (jridgewell) [inline src/dom.js]: `instanceof` has bug when crossing boundaries. Instead, use `n.nodeType == 9`

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 3]
```diff
@@ -186,6 +187,8 @@ export class AmpSlideScroll extends BaseSlides {
 
       // Slides must only be re-parented to DOM-connected nodes to avoid
       // errors when used in a shadow document (#9291).
+      dev().assert(isConnectedNode(slideWrapper),
```
Comment (jridgewell) [inline extensions/amp-carousel/0.1/slidescroll.js]: This'll probably affect everything that reparents.

File: src/dom.js (hunk 2 above)
Comment (camelburrito) [inline src/dom.js]: @jridgewell wow that is interesting! could you share if you have any reading material on this

Comment (jridgewell) [inline src/dom.js]: http://perfectionkills.com/instanceof-considered-harmful-or-how-to-write-a-robust-isarray/

Comment (choumx) [inline src/dom.js]: Good to know, though this shouldn't cross iframe boundary. Done.

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 3 above)
Comment (choumx) [inline extensions/amp-carousel/0.1/slidescroll.js]: Fair, created an issue to investigate: #9441

File: src/dom.js (hunk 1 above)
Comment (choumx) [inline src/dom.js]: I don't think it's any worse than `closestNode` etc. that walk up the ancestor chain.

**Comment by choumx** [comment]
@jridgewell @camelburrito Any other comments?

**Comment by choumx** [comment]
Thanks!

**Pull request merged**